File: SDK/C_SDK/tests/test_models.cpp
#include <tracehub/models.h>
#include <cassert>
#include <iostream>

void test_log_level_to_string() {
    assert(tracehub::log_level_to_string(tracehub::LogLevel::DEBUG) == "DEBUG");
    assert(tracehub::log_level_to_string(tracehub::LogLevel::INFO) == "INFO");
    assert(tracehub::log_level_to_string(tracehub::LogLevel::WARN) == "WARN");
    assert(tracehub::log_level_to_string(tracehub::LogLevel::ERROR) == "ERROR");
    assert(tracehub::log_level_to_string(tracehub::LogLevel::FATAL) == "FATAL");
    std::cout << "[PASS] test_log_level_to_string" << std::endl;
}

void test_string_to_log_level() {
    assert(tracehub::string_to_log_level("DEBUG") == tracehub::LogLevel::DEBUG);
    assert(tracehub::string_to_log_level("INFO") == tracehub::LogLevel::INFO);
    assert(tracehub::string_to_log_level("WARN") == tracehub::LogLevel::WARN);
    assert(tracehub::string_to_log_level("ERROR") == tracehub::LogLevel::ERROR);
    assert(tracehub::string_to_log_level("FATAL") == tracehub::LogLevel::FATAL);
    assert(tracehub::string_to_log_level("UNKNOWN") == tracehub::LogLevel::INFO);
    std::cout << "[PASS] test_string_to_log_level" << std::endl;
}

void test_log_entry_to_json() {
    tracehub::LogEntry entry;
    entry.level = tracehub::LogLevel::ERROR;
    entry.message = "Payment failed";
    entry.timestamp = "2026-04-05T14:30:00.123+00:00";
    entry.service = "billing-service";
    entry.environment = "production";
    entry.host = "server-1";
    entry.pid = 12345;
    entry.thread_id = "worker-1";
    entry.sdk_version = "1.2.0";
    entry.trace_id = "abc123";
    entry.module = "payments";
    entry.error_type = "ConnectionError";
    entry.error_message = "gateway timeout";
    entry.extra = {{"user_id", "usr_42"}};

    auto j = entry.to_json();

    assert(j["level"] == "ERROR");
    assert(j["message"] == "Payment failed");
    assert(j["service"] == "billing-service");
    assert(j["environment"] == "production");
    assert(j["host"] == "server-1");
    assert(j["pid"] == 12345);
    assert(j["trace_id"] == "abc123");
    assert(j["module"] == "payments");
    assert(j["error_type"] == "ConnectionError");
    assert(j["error_message"] == "gateway timeout");
    assert(j["extra"]["user_id"] == "usr_42");
    std::cout << "[PASS] test_log_entry_to_json" << std::endl;
}

void test_log_entry_roundtrip() {
    tracehub::LogEntry original;
    original.level = tracehub::LogLevel::WARN;
    original.message = "Disk full";
    original.timestamp = "2026-04-05T10:00:00.000+00:00";
    original.service = "infra";
    original.environment = "staging";
    original.host = "node-2";
    original.pid = 999;
    original.thread_id = "main";
    original.sdk_version = "1.2.0";

    auto j = original.to_json();
    auto restored = tracehub::LogEntry::from_json(j);

    assert(restored.level == original.level);
    assert(restored.message == original.message);
    assert(restored.service == original.service);
    assert(restored.pid == original.pid);
    assert(!restored.trace_id.has_value());
    std::cout << "[PASS] test_log_entry_roundtrip" << std::endl;
}

int main() {
    test_log_level_to_string();
    test_string_to_log_level();
    test_log_entry_to_json();
    test_log_entry_roundtrip();
    std::cout << "\nAll model tests passed!" << std::endl;
    return 0;
}

File: SDK/tests/test_models.py
"""Tests for sentinel_sdk.models."""

from sentinel_sdk.models import LogEntry, LogLevel


class TestLogLevel:
    def test_values(self):
        assert LogLevel.DEBUG == "DEBUG"
        assert LogLevel.INFO == "INFO"
        assert LogLevel.WARN == "WARN"
        assert LogLevel.ERROR == "ERROR"
        assert LogLevel.FATAL == "FATAL"

    def test_is_string_enum(self):
        assert isinstance(LogLevel.DEBUG, str)
        assert LogLevel.ERROR.value == "ERROR"


class TestLogEntry:
    def test_minimal_entry(self):
        entry = LogEntry(level=LogLevel.INFO, message="hello")
        assert entry.level == LogLevel.INFO
        assert entry.message == "hello"
        assert entry.module == ""
        assert entry.stack_trace is None
        assert entry.error_type is None
        assert entry.extra == {}

    def test_to_dict_minimal(self):
        entry = LogEntry(
            level=LogLevel.INFO,
            message="hello",
            timestamp="2026-04-05T10:00:00.000+00:00",
            service="test-svc",
            environment="test",
            host="localhost",
            pid=1234,
            thread_id="main",
            sdk_version="1.0.0",
        )
        d = entry.to_dict()
        assert d["level"] == "INFO"
        assert d["message"] == "hello"
        assert d["service"] == "test-svc"
        assert d["environment"] == "test"
        assert d["host"] == "localhost"
        assert d["pid"] == 1234
        assert "trace_id" not in d
        assert "module" not in d
        assert "stack_trace" not in d
        assert "error_type" not in d
        assert "extra" not in d

    def test_to_dict_full(self):
        entry = LogEntry(
            level=LogLevel.ERROR,
            message="fail",
            timestamp="2026-04-05T10:00:00.000+00:00",
            service="svc",
            environment="prod",
            host="host1",
            pid=99,
            thread_id="t1",
            sdk_version="1.0.0",
            trace_id="abc123",
            module="auth",
            stack_trace="Traceback...",
            error_type="ValueError",
            extra={"key": "val"},
        )
        d = entry.to_dict()
        assert d["trace_id"] == "abc123"
        assert d["module"] == "auth"
        assert d["stack_trace"] == "Traceback..."
        assert d["error_type"] == "ValueError"
        assert d["extra"] == {"key": "val"}
